impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -38,225 +82,312 @@ Join StrokePathGeometry::GetStrokeJoin() const {
-  vtx.position = position + start_offset * dir;
+  vtx = position + start_offset * dir;
   vtx_builder.AppendVertex(vtx);
-  vtx.position = position + end_offset * dir;
+  vtx = position + end_offset * dir;
   vtx_builder.AppendVertex(vtx);
 
   return dir;
 }
 
+// static
+void StrokePathGeometry::CreateMiterJoin(VertexWriter& vtx_builder,
+                                         const Point& position,
+                                         const Point& start_offset,
+                                         const Point& end_offset,
+                                         Scalar miter_limit,
+                                         Scalar scale) {
+  Point start_normal = start_offset.Normalize();
+  Point end_normal = end_offset.Normalize();
+
+  // 1 for no joint (straight line), 0 for max joint (180 degrees).
+  Scalar alignment = (start_normal.Dot(end_normal) + 1) / 2;
+  if (ScalarNearlyEqual(alignment, 1)) {
+    return;
+  }
+
+  Scalar dir = CreateBevelAndGetDirection(vtx_builder, position, start_offset,
+                                          end_offset);
+
+  Point miter_point = (start_offset + end_offset) / 2 / alignment;
+  if (miter_point.GetDistanceSquared({0, 0}) > miter_limit * miter_limit) {
+    return;  // Convert to bevel when we exceed the miter limit.
+  }
+
+  // Outer miter point.
+  VS::PerVertexData vtx;
+  vtx.position = position + miter_point * dir;
+  vtx_builder.AppendVertex(vtx.position);
+}
+// static
+void StrokePathGeometry::CreateRoundJoin(VertexWriter& vtx_builder,
+                                         const Point& position,
+                                         const Point& start_offset,
+                                         const Point& end_offset,
+                                         Scalar miter_limit,
+                                         Scalar scale) {
+  Point start_normal = start_offset.Normalize();
+  Point end_normal = end_offset.Normalize();
+
+  // 0 for no joint (straight line), 1 for max joint (180 degrees).
+  Scalar alignment = 1 - (start_normal.Dot(end_normal) + 1) / 2;
+  if (ScalarNearlyEqual(alignment, 0)) {
+    return;
+  }
+
+  Scalar dir = CreateBevelAndGetDirection(vtx_builder, position, start_offset,
+                                          end_offset);
+
+  Point middle =
+      (start_offset + end_offset).Normalize() * start_offset.GetLength();
+  Point middle_normal = middle.Normalize();
+
+  Point middle_handle = middle + Point(-middle.y, middle.x) *
+                                     PathBuilder::kArcApproximationMagic *
+                                     alignment * dir;
+  Point start_handle = start_offset + Point(start_offset.y, -start_offset.x) *
+                                          PathBuilder::kArcApproximationMagic *
+                                          alignment * dir;
+
+  VS::PerVertexData vtx;
+  CubicPathComponent(start_offset, start_handle, middle_handle, middle)
```

Comment:
We talked about this offline - should we have some specialized codepath for this since we know we're trying to approximate a circle? If not for this patch, for a follow up bug?

Reply:
Filled https://github.com/flutter/flutter/issues/143012 .